src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkTransition.java

```diff
@@ -329,11 +329,20 @@ private static Target getActual(
         if (allowsMultiple) {
           // if this setting allows multiple settings
           if (!(newValue instanceof List)) {
-            throw new TransitionException(
-                String.format(
-                    "'%s' allows multiple values and must be set"
-                        + " in transition using a starlark list instead of single value '%s'",
-                    actualSetting, newValue));
+            if (newValue.equals(rule.getAttr(STARLARK_BUILD_SETTING_DEFAULT_ATTR_NAME))) {
```

Comment:
It really took me some time to realize `changedSettingToRule` from line 310 is now a misleading name.  As I understand, the reality is:

1. `changedSettingToRule` was originally written to encapsulate settings the transition changed
2. `build_setting` inputs are implicitly changed by transition code as an implementation detail, from non-existent to their explicit default values, to guarantee the transition can read them
3. https://github.com/bazelbuild/bazel/pull/13971 causes the original logic, which historically only checked user-changed settings, to also include these implicit changes. For the purpose of guaranteeing the default that the transition added are then removed.

I'm worried about these overloading of concepts, even if it keeps the code concise.

What if we broke out the logic some: revert https://github.com/bazelbuild/bazel/pull/13971/files, then do a separate distinct call to `getRelevantStarlarkSettingsFromTransition(Settings.INPUTS)` which *only* replicates the part of `validate` that removes an option if it's at its default value?

I realize that takes more refactoring, likely breaking out a helper method for `validate()`. But I'd really like to keep the code as clear as possible in transition logic, since it's super-complicated as it is.

At the very leas I'd like the naming and commenting (including method Javadoc) to remain accurate. I worry it could be more work to clearly make the commenting accurate than do the refactoring I'm suggesting.

Reply:
That's a very good point - input build settings should be cleared unconditionally rather than undergo validation. I created https://github.com/bazelbuild/bazel/pull/14972 for this refactoring and a better fix for #14894.